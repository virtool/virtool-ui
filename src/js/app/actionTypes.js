/**
 * Create a special action type used for requests.
 *
 * The request is an object with three properties assigned with action types used for API requests.
 *
 * - REQUESTED
 * - SUCCEEDED
 * - FAILED
 *
 * @param root {string} the root name of the action type
 * @returns {object} a request-style action type
 */
const createRequestActionType = root => ({
    REQUESTED: `${root}_REQUESTED`,
    SUCCEEDED: `${root}_SUCCEEDED`,
    FAILED: `${root}_FAILED`,
});

// App
export const GET_INITIAL_STATE = createRequestActionType("GET_INITIAL_STATE");

// Account
export const GET_ACCOUNT = createRequestActionType("GET_ACCOUNT");
export const UPDATE_ACCOUNT = createRequestActionType("UPDATE_ACCOUNT");
export const CHANGE_ACCOUNT_PASSWORD = createRequestActionType("CHANGE_ACCOUNT_PASSWORD");
export const LOGIN = createRequestActionType("LOGIN");
export const LOGOUT = createRequestActionType("LOGOUT");
export const RESET_PASSWORD = createRequestActionType("RESET_PASSWORD");

// Administration Settings
export const GET_SETTINGS = createRequestActionType("GET_SETTINGS");
export const UPDATE_SETTINGS = createRequestActionType("UPDATE_SETTINGS");

// Analysis
export const WS_UPDATE_ANALYSIS = "WS_UPDATE_ANALYSIS";
export const SET_ANALYSIS = "SET_ANALYSIS";
export const SET_ANALYSIS_ACTIVE_ID = "SET_ANALYSIS_ACTIVE_ID";
export const SET_SEARCH_IDS = "SET_SEARCH_IDS";
export const SET_AODP_FILTER = "ADD_AODP_FILTER";
export const SET_ANALYSIS_SORT_KEY = "SET_ANALYSIS_SORT_KEY";

// Errors
export const CLEAR_ERROR = "CLEAR_ERROR";

// Files
export const UPLOAD = createRequestActionType("UPLOAD");
export const UPLOAD_SAMPLE_FILE = createRequestActionType("UPLOAD_SAMPLE_FILE");
export const UPLOAD_PROGRESS = "UPLOAD_PROGRESS";
export const UPLOAD_FAILED = "UPLOAD_FAILED";
export const REMOVE_UPLOAD = "REMOVE_UPLOAD";

// Forms
export const SET_PERSISTENT_FORM_STATE = "SET_PERSISTENT_FORM_STATE";
export const DELETE_PERSISTENT_FORM_STATE = "DELETE_PERSISTENT_FORM_STATE";
export const SET_REDUX_FORM_STATE = "SET_REDUX_FORM_STATE";

// Groups
export const CREATE_GROUP = createRequestActionType("CREATE_GROUP");

// HMMs
export const GET_HMM = createRequestActionType("GET_HMM");

// Indexes
export const GET_INDEX = createRequestActionType("GET_INDEX");
export const CREATE_INDEX = createRequestActionType("CREATE_INDEX");

// Jobs
export const GET_JOB = createRequestActionType("GET_JOB");

// OTU
export const GET_OTU = createRequestActionType("GET_OTU");
export const CREATE_OTU = createRequestActionType("CREATE_OTU");
export const EDIT_OTU = createRequestActionType("EDIT_OTU");
export const ADD_ISOLATE = createRequestActionType("ADD_ISOLATE");
export const EDIT_ISOLATE = createRequestActionType("EDIT_ISOLATE");
export const ADD_SEQUENCE = createRequestActionType("ADD_SEQUENCE");
export const EDIT_SEQUENCE = createRequestActionType("EDIT_SEQUENCE");

// Refs
export const GET_REFERENCE = createRequestActionType("GET_REFERENCE");

// Samples
export const GET_SAMPLE = createRequestActionType("GET_SAMPLE");
export const CREATE_SAMPLE = createRequestActionType("CREATE_SAMPLE");
export const UPDATE_SAMPLE = createRequestActionType("UPDATE_SAMPLE");

// Subtraction
export const GET_SUBTRACTION = createRequestActionType("GET_SUBTRACTION");
export const CREATE_SUBTRACTION = createRequestActionType("CREATE_SUBTRACTION");

// Users
export const FIND_USERS = createRequestActionType("FIND_USERS");
export const GET_USER = createRequestActionType("GET_USER");
export const CREATE_USER = createRequestActionType("CREATE_USER");
export const CREATE_FIRST_USER = createRequestActionType("CREATE_FIRST_USER");
export const EDIT_USER = createRequestActionType("EDIT_USER");
