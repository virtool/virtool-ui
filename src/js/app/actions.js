import { createAction } from "@reduxjs/toolkit";
import { GET_INITIAL_STATE } from "./actionTypes";

/**
 * Returns action that can trigger an API calls to get the initial state of the application
 *
 * @func
 * @returns {object}
 */
export const getInitialState = createAction(GET_INITIAL_STATE.REQUESTED);
