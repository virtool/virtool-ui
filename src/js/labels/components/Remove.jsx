import { DialogPortal, DialogTrigger } from "@radix-ui/react-dialog";
import React, { useState } from "react";
import { useMutation, useQueryClient } from "react-query";
import styled from "styled-components";
import { Request } from "../../app/request";
import { getFontSize } from "../../app/theme";
import { Button, Dialog, DialogContent, DialogOverlay, DialogTitle } from "../../base";
import { StyledButtonSmall } from "../../base/styled/StyledButtonSmall";

const RemoveLabelQuestion = styled.p`
    font-size: ${getFontSize("lg")};
`;

const RemoveLabelFooter = styled.footer`
    display: flex;
    margin-top: 30px;
`;

export function RemoveLabel({ id, name }) {
    const [show, setShow] = useState(false);

    const queryClient = useQueryClient();

    const mutation = useMutation(() => {
        return Request.delete(`/api/labels/${id}`);
    });

    const handleDelete = () => {
        mutation.mutate(["label", id], {
            onSuccess: () => {
                setShow(false);
                queryClient.invalidateQueries("labels");
            }
        });
    };

    return (
        <Dialog show={show} onOpenChange={open => setShow(open)}>
            <StyledButtonSmall as={DialogTrigger}>Delete</StyledButtonSmall>
            <DialogPortal>
                <DialogOverlay />
                <DialogContent>
                    <DialogTitle>Delete Label</DialogTitle>
                    <RemoveLabelQuestion>
                        Are you sure you want to delete the label <strong>{name}</strong>?
                    </RemoveLabelQuestion>
                    <RemoveLabelFooter>
                        <Button type="button" color="red" onClick={handleDelete}>
                            Delete
                        </Button>
                    </RemoveLabelFooter>
                </DialogContent>
            </DialogPortal>
        </Dialog>
    );
}
