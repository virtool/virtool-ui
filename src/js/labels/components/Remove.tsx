import { DialogPortal, DialogTrigger } from "@radix-ui/react-dialog";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import React, { useState } from "react";
import styled from "styled-components";
import { Request } from "../../app/request";
import { getFontSize } from "../../app/theme";
import { Button, Dialog, DialogContent, DialogOverlay, DialogTitle } from "../../base";
import { StyledButtonSmall } from "../../base/styled/StyledButtonSmall";
import { labelQueryKeys } from "../queries";

const RemoveLabelQuestion = styled.p`
    font-size: ${getFontSize("lg")};
`;

const RemoveLabelFooter = styled.footer`
    display: flex;
    margin-top: 30px;
`;

interface RemoveLabelProps {
    id: string;
    name: string;
}

export function RemoveLabel({ id, name }: RemoveLabelProps) {
    const [open, setOpen] = useState(false);

    const queryClient = useQueryClient();

    const mutation = useMutation(
        () => {
            return Request.delete(`/labels/${id}`);
        },
        {
            onSuccess: () => {
                setOpen(false);
                queryClient.invalidateQueries(labelQueryKeys.lists());
            },
        },
    );

    return (
        <Dialog open={open} onOpenChange={open => setOpen(open)}>
            <StyledButtonSmall as={DialogTrigger}>Delete</StyledButtonSmall>
            <DialogPortal>
                <DialogOverlay />
                <DialogContent>
                    <DialogTitle>Delete Label</DialogTitle>
                    <RemoveLabelQuestion>
                        Are you sure you want to delete the label <strong>{name}</strong>?
                    </RemoveLabelQuestion>
                    <RemoveLabelFooter>
                        <Button type="button" color="red" onClick={mutation.mutate}>
                            Delete
                        </Button>
                    </RemoveLabelFooter>
                </DialogContent>
            </DialogPortal>
        </Dialog>
    );
}
