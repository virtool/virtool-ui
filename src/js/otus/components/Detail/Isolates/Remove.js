import { get } from "lodash-es";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import { RemoveModal } from "../../../../base";
import { hideOTUModal, removeIsolate } from "../../../actions";

export const Remove = ({ id, name, nextId, otuId, show, onConfirm, onHide }) => {
    const handleConfirm = useCallback(() => {
        onConfirm(otuId, id, nextId);
    }, [otuId, id]);

    return <RemoveModal name={name} noun="Isolate" onConfirm={handleConfirm} onHide={onHide} show={show} />;
};

export const mapStateToProps = state => {
    const { id, name } = state.otus.activeIsolate;
    return {
        id,
        name,
        nextId: get(state, ["otus", "detail", "isolates", 0, "id"], null),
        otuId: state.otus.detail.id,
        show: state.otus.removeIsolate
    };
};

export const mapDispatchToProps = dispatch => ({
    onHide: () => {
        dispatch(hideOTUModal());
    },

    onConfirm: (otuId, isolateId, nextIsolateId) => {
        dispatch(removeIsolate(otuId, isolateId, nextIsolateId));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Remove);
