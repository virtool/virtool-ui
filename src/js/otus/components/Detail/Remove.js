import React, { useCallback } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { removeOTU, hideOTUModal } from "../../actions";
import { RemoveModal } from "../../../base";

export const Remove = ({ history, id, name, refId, show, onConfirm, onHide }) => {
    const handleConfirm = useCallback(() => {
        onConfirm(refId, id, history);
    }, [id, refId]);

    return <RemoveModal name={name} noun="OTU" onConfirm={handleConfirm} onHide={onHide} show={show} />;
};

Remove.propTypes = {
    history: PropTypes.object,
    id: PropTypes.string,
    name: PropTypes.string,
    refId: PropTypes.string,
    show: PropTypes.bool,
    onConfirm: PropTypes.func,
    onHide: PropTypes.func
};

export const mapStateToProps = state => ({
    show: state.otus.remove,
    refId: state.references.detail.id
});

export const mapDispatchToProps = dispatch => ({
    onConfirm: (refId, otuId, history) => {
        dispatch(removeOTU(refId, otuId, history));
    },

    onHide: () => {
        dispatch(hideOTUModal());
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Remove);
