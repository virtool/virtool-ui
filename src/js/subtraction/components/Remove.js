import React from "react";
import { connect } from "react-redux";
import { pushState } from "../../app/actions";
import { RemoveModal } from "../../base";
import { routerLocationHasState } from "../../utils/utils";
import { removeSubtraction } from "../actions";

export const RemoveSubtraction = ({ id, name, show, onConfirm, onHide }) => (
    <RemoveModal id={id} name={name} noun="Subtraction" show={show} onHide={onHide} onConfirm={() => onConfirm(id)} />
);

export const mapStateToProps = state => ({
    id: state.subtraction.detail.id,
    name: state.subtraction.detail.name,
    show: routerLocationHasState(state, "removeSubtraction", true)
});

export const mapDispatchToProps = dispatch => ({
    onHide: () => {
        dispatch(pushState({ removeSubtraction: false }));
    },

    onConfirm: subtractionId => {
        dispatch(removeSubtraction(subtractionId));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(RemoveSubtraction);
