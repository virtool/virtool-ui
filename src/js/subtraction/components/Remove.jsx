import React from "react";
import { connect } from "react-redux";
import { pushState } from "../../app/actions";
import { RemoveModal } from "../../base";
import { routerLocationHasState } from "../../utils/utils";
import { removeSubtraction } from "../actions";

export const RemoveSubtraction = ({ subtraction, show, onConfirm, onHide }) => (
    <RemoveModal
        id={subtraction.id}
        name={subtraction.name}
        noun="Subtraction"
        show={show}
        onHide={onHide}
        onConfirm={() => onConfirm(subtraction.id)}
    />
);

export const mapStateToProps = state => ({
    show: routerLocationHasState(state, "removeSubtraction", true),
});

export const mapDispatchToProps = dispatch => ({
    onHide: () => {
        dispatch(pushState({ removeSubtraction: false }));
    },

    onConfirm: subtractionId => {
        dispatch(removeSubtraction(subtractionId));
    },
});

export default connect(mapStateToProps, mapDispatchToProps)(RemoveSubtraction);
