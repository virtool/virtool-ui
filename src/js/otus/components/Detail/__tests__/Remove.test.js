import { HIDE_OTU_MODAL, REMOVE_OTU } from "../../../../app/actionTypes";
import { Remove, mapStateToProps, mapDispatchToProps } from "../Remove";

describe("<Remove />", () => {
    let props;

    beforeEach(() => {
        props = {
            history: {},
            id: "foo",
            name: "Foo",
            refId: "baz",
            show: true,
            onConfirm: jest.fn(),
            onHide: jest.fn()
        };
    });

    it("should render when [show=true]", () => {
        const wrapper = shallow(<Remove {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [show=false]", () => {
        props.show = false;
        const wrapper = shallow(<Remove {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onConfirm() when onConfirm() called on <RemoveModal />", () => {
        const wrapper = shallow(<Remove {...props} />);
        wrapper.props().onConfirm();
        expect(props.onConfirm).toHaveBeenCalledWith("baz", "foo", props.history);
    });

    it("should call onHide() when onHide() called on <RemoveModal />", () => {
        const wrapper = shallow(<Remove {...props} />);
        wrapper.props().onHide();
        expect(props.onHide).toHaveBeenCalled();
    });
});

describe("mapStateToProps()", () => {
    let state;

    beforeEach(() => {
        state = {
            otus: {
                remove: false
            },
            references: {
                detail: {
                    id: "foo"
                }
            }
        };
    });

    it.each([true, false])("should return props when [state.otus.remove=%p]", show => {
        state.otus.remove = show;
        const props = mapStateToProps(state);
        expect(props).toEqual({
            show,
            refId: "foo"
        });
    });
});

describe("mapDispatchToProps()", () => {
    let dispatch;
    let props;

    beforeEach(() => {
        dispatch = jest.fn();
        props = mapDispatchToProps(dispatch);
    });

    it("should return onConfirm() in props", () => {
        const refId = "foo";
        const otuId = "baz";
        const history = {
            id: "bar"
        };
        props.onConfirm(refId, otuId, history);
        expect(dispatch).toHaveBeenCalledWith({
            type: REMOVE_OTU.REQUESTED,
            history,
            otuId,
            refId
        });
    });

    it("should return onHide() in props", () => {
        props.onHide();
        expect(dispatch).toHaveBeenCalledWith({
            type: HIDE_OTU_MODAL
        });
    });
});
