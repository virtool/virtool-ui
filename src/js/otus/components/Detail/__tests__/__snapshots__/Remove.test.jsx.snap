// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Remove /> > should render when [show=false] 1`] = `
<RemoveModal
  name="Foo"
  noun="OTU"
  onConfirm={[Function]}
  onHide={[MockFunction spy]}
  show={false}
/>
`;

exports[`<Remove /> > should render when [show=true] 1`] = `
<RemoveModal
  name="Foo"
  noun="OTU"
  onConfirm={[Function]}
  onHide={[MockFunction spy]}
  show={true}
/>
`;
